{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# LLM week 3 Homework \n",
    "\n",
    "-- by Junjie Zhang"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Assignment: Implement a RAG-based Chatbot using Langchain\n",
    "\n",
    "\n",
    "-   • A console program to continuously receive user input and respond\n",
    "-   • Try smaller dataset first and then larger dataset\n",
    "-   • Refer to the examples in\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Summary\n",
    "\n",
    "I loaded 11 literatures from the local disk using a directory loader for RAG and chatbot. THe contents are about genetic breeding which is used for seeds innovation. \n",
    "\n",
    "**This homework includes the following steps:**\n",
    "1. **Environment Setup**\n",
    "    - Import necessary libraries and modules.\n",
    "    - Load environment variables from `.env` file to get the API key.\n",
    "\n",
    "2. **Document Processing**\n",
    "    - Extract text from multiple PDF files and save each to a separate text file.\n",
    "    - Load documents from the local disk using a directory loader.\n",
    "    - Split documents into smaller chunks for better processing.\n",
    "\n",
    "3. **Embedding Creation**\n",
    "    - Create embeddings for each chunk using OpenAI embeddings.\n",
    "    - Create a retriever from the vector store to fetch relevant chunks.\n",
    "\n",
    "4. **Conversational Chain Setup**\n",
    "    - Initialize the chat model.\n",
    "    - Create a memory object to store the conversation history.\n",
    "    - Create a conversational retrieval chain by combining the chat model, retriever, and memory.\n",
    "\n",
    "5. **User Interaction**\n",
    "    - Start an infinite loop to interact with the user.\n",
    "    - Continuously receive user input and respond using the conversational retrieval chain.\n",
    "    - Exit the loop if the user inputs \"exit\".\n",
    "\n",
    "**Challenges:**\n",
    "1. **Document Processing:** I had difficulty in reading the PDF files by DirectoryLoader. \n",
    "2. **performance:** The performance of the chatbot is not good enough. It can retrieved the relevant information but it is not deep enough. May need more data and better model to improve the performance.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "from openai import OpenAI\n",
    "from dotenv import load_dotenv\n",
    "import os\n",
    "\n",
    "# Load environment variables from .env file\n",
    "load_dotenv()\n",
    "\n",
    "# Get the API key\n",
    "LANGSMITH_API_KEY = os.getenv('LANGSMITH_API_KEY')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from langchain.chains import ConversationalRetrievalChain\n",
    "from langchain_community.vectorstores import Chroma\n",
    "from langchain_openai import ChatOpenAI, OpenAIEmbeddings\n",
    "from langchain.memory import ConversationBufferMemory, ConversationSummaryMemory\n",
    "from langchain_community.document_loaders import DirectoryLoader\n",
    "from langchain.text_splitter import RecursiveCharacterTextSplitter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/Users/junjiezhang/Documents/GitHub/LLM_Zeng24/homework'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import glob\n",
    "# import PyPDF2\n",
    "\n",
    "# # Extract the text from multiple PDF files and write each to a separate text file\n",
    "# def extract_text_from_pdfs():\n",
    "#     context_files = glob.glob(\"../files/genetics_breeding/*.pdf\")\n",
    "#     for file_path in context_files:\n",
    "#         output_file_path = file_path.replace(\".pdf\", \".txt\")\n",
    "#         with open(file_path, 'rb') as file:\n",
    "#             reader = PyPDF2.PdfReader(file)\n",
    "#             text = \"\"\n",
    "#             for page_num in range(len(reader.pages)):\n",
    "#                 page = reader.pages[page_num]\n",
    "#                 text += page.extract_text() + \"\\n\"\n",
    "#             with open(output_file_path, 'w') as txt_file:\n",
    "#                 txt_file.write(text)\n",
    "\n",
    "# extract_text_from_pdfs()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#pip install --upgrade pypdf langchain"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Ignoring wrong pointing object 2588 0 (offset 0)\n",
      "Ignoring wrong pointing object 2592 0 (offset 0)\n",
      "Ignoring wrong pointing object 4257 0 (offset 0)\n",
      "Ignoring wrong pointing object 4258 0 (offset 0)\n",
      "Ignoring wrong pointing object 4259 0 (offset 0)\n",
      "Ignoring wrong pointing object 4260 0 (offset 0)\n",
      "Ignoring wrong pointing object 4261 0 (offset 0)\n",
      "Ignoring wrong pointing object 4262 0 (offset 0)\n",
      "Ignoring wrong pointing object 74 0 (offset 0)\n",
      "Ignoring wrong pointing object 76 0 (offset 0)\n"
     ]
    }
   ],
   "source": [
    "\n",
    "from langchain_community.document_loaders import DirectoryLoader\n",
    "from langchain_community.document_loaders import PyPDFDirectoryLoader\n",
    "# Load all PDF documents from the local directory\n",
    "loader = PyPDFDirectoryLoader('../files/genetics_breeding', glob=\"*.pdf\")\n",
    "documents = loader.load()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Document(metadata={'source': '../files/genetics_breeding/Agronomy Journal - 2015 - Casler - Fundamentals of Experimental Design  Guidelines for Designing Successful Experiments.pdf', 'page': 0}, page_content='692\\t Agronomy\\tJournal\\t •\\t Volume\\t107,\\tIssue\\t2\\t •\\t 2015\\nSymposium: Statistical Concepts\\nFundamentals\\tof\\tExperimental\\tDesign:\\tGuidelines\\tfor\\t\\nDesigning\\tSuccessful\\tExperiments\\nMichael\\tD.\\tCasler*\\nPublished in Agron. J. 107:692–705 (2015)\\ndoi:10.2134/agronj2013.0114\\nAvailable freely online through the author-supported open access option.\\nCopyright © 2015 by the American Society of Agronomy, 5585 Guilford \\nRoad, Madison, WI 53711. All rights reserved. No part of this periodical may \\nbe reproduced or transmitted in any form or by any means, electronic or \\nmechanical, including photocopying, recording, or any information storage \\nand retrieval system, without permission in writing from the publisher.\\nDesigning\\texperiments\\tinvolves\\ta\\tmarriage of biologi-\\ncal and mathematical sciences. Th  e mathematical, or statistical, \\nscience is obvious. We use scientifi  c fundamentals and principles \\nthat have been developed during the past century to conduct \\nthree types of experiments. Observational experiments are \\nthose designed to measure or verify an assumed constant, such \\nas the velocity of light or the mass of an atom. Measurement \\nexperiments are those designed to measure the properties of a \\npopulation, the members of which are variable, such as com-\\nmodity prices, production statistics, or neutrino frequencies. As \\nbiologists, we are principally concerned with comparative (or \\nmanipulative) experiments, in which our global goal is to com-\\npare or contrast two or more practices or systems that may have \\nsome relevance to our fi  eld of scientifi  c inquiry. It is the solutions, \\nor the specifi  c choices we must make, that are not so obvious.\\nIn conducting comparative experiments, we routinely fol-\\nlow the general theory of scientifi  c inquiry shown in Fig. 1. We \\nbegin with questions and/or hypotheses that must be trans-\\nlated to models based on the specifi  c subject matter, e.g., cotton \\n(Gossypium hirsutum L.) plants, farm machinery, or hay bales. \\nTh  e subject matter model is translated into a statistical model, \\nwhich is developed in concert with the statistical design. Th  e \\nstatistical design includes both the treatment design and the \\nexperimental design and provides a set of rules and procedures \\nthat allow us to conduct the experiment. For many of us, this \\nprocess becomes routine, such that we tend to forget the fun-\\ndamental nature and assumptions of statistical designs, instead \\nforming designs and forging ahead using time-honored and \\ntraditional approaches that have worked well for us in the past. \\nWe oft  en favor familiarity, simplicity, and constancy over any \\nthought of change or concept of improvement. Researchers \\nmust recognize that designing comparative experiments is a \\nmassively decision-based exercise, so that the truism “If you \\nchoose not to decide, you still have made a choice” (Peart, \\n1980) is an appropriate concept in biological research.\\nOnce the experiment has been conducted and the data col-\\nlected, the statistical analysis proceeds as determined by the \\nresearcher before the experiment was conducted. Th  e analysis \\nleads to specifi  c interpretations of the results, creating inferences \\nand conclusions that bring the research back to the original \\nquestion or hypothesis. Finally, and a point oft  en ignored in \\nmost models, is the feedback loop that comes when the experi-\\nment is completed (Fig. 1). As scientists, we are intimately famil-\\niar with this feedback loop, oft  en correctly opining that, “Any \\ngood experiment leads to more questions and new hypotheses \\nthan it answers.” What we tend to forget or ignore is the fact \\nthat two types of information travel along this feedback loop: \\nscientifi  c answers to our biological questions and mathematical \\nor statistical information that can be used to design better, more \\neffi   cient, future experiments. Just as any experiment helps the \\nABSTRACT\\nWe o/f_t  en think of experimental designs as analogous to recipes in a cookbook. We look for something that we like, something that \\nsatis/f_i  es our needs, and frequently return to those that have become our long-standing favorites. We can easily become complacent, \\nfavoring the tried-and-true designs (or recipes) over those that contain unknown or untried ingredients or those that are too complex \\nfor our tastes and skills. Instead, I prefer to think of experimental designs as a creative series of decisions that are meant to solve one \\nor more problems. /T_h  ese problems may be real or imagined—we may have direct evidence of a past or current problem or we may \\nsimply want insurance against future potential problems. /T_h  e most signi/f_i  cant manifestation of a “problem” or a “failed” design \\nis unsatisfactory P values that prevent us from developing inferences about treatment diﬀ  erences. Four basic tenets or pillars of \\nexperimental design— replication, randomization, blocking, and size of experimental units— can be used creatively, intelligently, \\nand consciously to solve both real and perceived problems in comparative experiments. Because research is expensive, both in terms \\nof grant funds and the emotional costs invested in grant competition and administration, biological experiments should be designed \\nunder the mantra “failure is not an option.” Guidelines and advice provided in this review are designed to reduce the probability of \\nfailure for researchers who are willing to question, evaluate, and possibly modify their decision-making processes.\\nUSDA-ARS, U.S. Dairy Forage Research Center, 1925 Linden Dr. W., \\nMadison, WI  53706. Received 5 Mar. 2013. Accepted 4 May 2013. \\n*Corresponding author (michaelcasler@ars.usda.gov).\\nAbbreviations:  CRD, completely randomized design; RCBD, randomized \\ncomplete block design.Published February 19, 2014\\n 14350645, 2015, 2, Downloaded from https://acsess.onlinelibrary.wiley.com/doi/10.2134/agronj2013.0114 by Test, Wiley Online Library on [06/02/2023]. See the Terms and Conditions (https://onlinelibrary.wiley.com/terms-and-conditions) on Wiley Online Library for rules of use; OA articles are governed by the applicable Creative Commons License\\n')"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "documents[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# some pdf files have problem to load, use this code to load one by one to identify the problem files. \n",
    "# import os\n",
    "# from langchain.document_loaders import PyPDFLoader\n",
    "\n",
    "# directory = '../files/genetics_breeding'\n",
    "\n",
    "# # List all PDF files in the directory\n",
    "# pdf_files = [f for f in os.listdir(directory) if f.endswith('.pdf')]\n",
    "\n",
    "# # Check each PDF file\n",
    "# for pdf_file in pdf_files:\n",
    "#     try:\n",
    "#         pdf_path = os.path.join(directory, pdf_file)\n",
    "#         loader = PyPDFLoader(pdf_path)\n",
    "#         documents = loader.load()\n",
    "#         print(f\"{pdf_file}: Loaded successfully with {len(documents)} documents.\")\n",
    "#     except Exception as e:\n",
    "#         print(f\"{pdf_file}: Failed to load. Error: {e}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# print(\"Loading documents\")\n",
    "# # Load documents from the local disk using a directory loader\n",
    "# loader = DirectoryLoader('../files/genetics_breeding', glob=\"*.txt\")\n",
    "# documents = loader.load()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Creating embeddings\n",
      "Creating chains\n"
     ]
    }
   ],
   "source": [
    "# Split documents into smaller chunks for better processing\n",
    "text_splitter = RecursiveCharacterTextSplitter(chunk_size=1000, chunk_overlap=100)\n",
    "chunks = text_splitter.split_documents(documents)\n",
    "\n",
    "print(\"Creating embeddings\")\n",
    "\n",
    "# Create embeddings for each chunk using OpenAI embeddings\n",
    "embeddings = OpenAIEmbeddings()\n",
    "vectorstore = Chroma.from_documents(chunks, embeddings)\n",
    "\n",
    "# Create a retriever from the vector store to fetch relevant chunks\n",
    "retriever = vectorstore.as_retriever()\n",
    "\n",
    "print(\"Creating chains\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "ii\n",
      "\n",
      "●\n",
      "●\n",
      "●\n",
      "●\n",
      "●\n",
      "●●● ●●\n",
      "●\n",
      "● ●● ● ●\n",
      "●\n",
      "●\n",
      "●\n",
      "●\n",
      "●●\n",
      "● ●● ●\n",
      "●\n",
      "● ●●\n",
      "●●\n",
      "●●\n",
      "●\n",
      "●\n",
      "●\n",
      "●\n",
      "●\n",
      "●\n",
      "●\n",
      "●●\n",
      "●\n",
      "●\n",
      "●\n",
      "●\n",
      "●\n",
      "●\n",
      "●● ●● ● ●\n",
      "●● ● ●\n",
      "●●●\n",
      "●\n",
      "●\n",
      "●● ●\n",
      "●● ●\n",
      "● ●\n",
      "●●\n",
      "●● ●● ● ●\n",
      "●\n",
      "●●●\n",
      "● ●\n",
      "●\n",
      "●●\n",
      "●\n",
      "●\n",
      "●\n",
      "●\n",
      "●\n",
      "●\n",
      "●\n",
      "●\n",
      "●\n",
      "●●\n",
      "● ●\n",
      "●\n",
      "●●\n",
      "●● ●●●\n",
      "●●\n",
      "●●\n",
      "●\n",
      "●\n",
      "●\n",
      "●\n",
      "●●\n",
      "●\n",
      "● ●\n",
      "●\n",
      "●\n",
      "●\n",
      "●\n",
      "●\n",
      "●\n",
      "●\n",
      "● ●\n",
      "●\n",
      "●\n",
      "● ●\n",
      "● ●\n",
      "●\n",
      "●\n",
      "●\n",
      "●\n",
      "●\n",
      "●●\n",
      "●\n",
      "●\n",
      "●●\n",
      "●\n",
      "●\n",
      "●\n",
      "●\n",
      "● ●\n",
      "● ●●\n",
      "●\n",
      "●● ●\n",
      "●\n",
      "●\n",
      "● ●\n",
      "●\n",
      "●\n",
      "●\n",
      "●● ● ●\n",
      "●●\n",
      "●\n",
      "●● ●\n",
      "● ●●\n",
      "●\n",
      "●\n",
      "●\n",
      "●\n",
      "● ●\n",
      "●● ●●●\n",
      "●\n",
      "● ●\n",
      "●\n",
      "●●\n",
      "●\n",
      "●● ●\n",
      "●●\n",
      "●\n",
      "●\n",
      "● ●\n",
      "●●\n",
      "●●●\n",
      "●\n",
      "●\n",
      "● ●●\n",
      "● ●●\n",
      "●\n",
      "●\n",
      "●\n",
      "● ●\n",
      "●\n",
      "●\n",
      "●\n",
      "●\n",
      "●\n",
      "●●\n",
      "●\n",
      "●\n",
      "●● ●●● ●\n",
      "●●\n",
      "● ●\n",
      "●\n",
      "●\n",
      "●\n",
      "●\n",
      "●\n",
      "●\n",
      "●●\n",
      "●\n",
      "● ●\n",
      "●●●\n",
      "●\n",
      "● ●\n",
      "●●\n",
      "●●\n",
      "● ● ●\n",
      "●\n",
      "●\n",
      "●\n",
      "●\n",
      "●\n",
      "●\n",
      "● ● ●\n",
      "●●\n",
      "●\n",
      "●●\n",
      "●\n",
      "●\n",
      "●\n",
      "●\n",
      "● ●\n",
      "●● ●\n",
      "●●\n",
      "●\n",
      "●\n",
      "●\n",
      "●●\n",
      "●\n",
      "●\n",
      "● ●●\n",
      "●\n",
      "● ●\n",
      "●● ●\n",
      "●●\n",
      "●\n",
      "●\n",
      "●\n",
      "●\n",
      "●\n",
      "●\n",
      "●\n",
      "●\n",
      "●●● ●\n",
      "●\n",
      "●\n",
      "●●\n",
      "● ●\n",
      "●\n",
      "●●\n",
      "● ●● ●\n",
      "●\n",
      "●●\n",
      "●\n",
      "●\n",
      "●\n",
      "●● ● ●\n",
      "●\n",
      "●\n",
      "●\n",
      "●\n",
      "●\n",
      "● ●\n",
      "●\n",
      "●\n",
      "●\n",
      "●\n",
      "●\n",
      "● ●\n",
      "●\n",
      "● ●● ●\n",
      "●\n",
      "●\n",
      "●\n",
      "●\n",
      "●\n",
      "● ●\n",
      "●\n",
      "●\n",
      "● ●●●\n",
      "●\n",
      "●\n",
      "●\n",
      "●\n",
      "●\n",
      "●\n",
      "●\n",
      "●● ● ●\n",
      "●● ●\n",
      "●\n",
      "●\n",
      "●\n",
      "●\n",
      "●\n",
      "●\n",
      "● ●\n",
      "● ●\n",
      "●\n",
      "●\n",
      "●\n",
      "●\n",
      "●\n",
      "●● ●\n",
      "●●●\n",
      "●\n",
      "●\n",
      "● ●\n",
      "● ●\n",
      "●\n",
      "●\n",
      "●\n",
      "●●\n",
      "●\n",
      "● ●\n",
      "●\n",
      "●\n",
      "● ●●\n",
      "● ●\n",
      "●\n",
      "●\n",
      "● ●●\n",
      "●\n",
      "●\n",
      "●\n",
      "● ● ●\n",
      "●\n",
      "●\n",
      "●\n",
      "● ●\n",
      "●\n",
      "●\n",
      "● ●\n",
      "●\n",
      "●\n",
      "●\n",
      "●\n",
      "●\n",
      "●\n",
      "●\n",
      "● ●●\n",
      "●\n",
      "●● ●\n",
      "●\n",
      "●\n",
      "●●\n",
      "●\n",
      "●\n",
      "●\n",
      "●\n",
      "●\n",
      "● ●\n",
      "●\n",
      "●\n",
      "●\n",
      "●\n",
      "●●● ●\n",
      "●●● ●\n",
      "● ●●\n",
      "●\n",
      "●●●\n",
      "●●\n",
      "●\n",
      "1\n",
      "100\n",
      "150 300\n",
      "1000\n",
      "1500\n",
      "2000\n",
      "2500\n",
      "\n",
      "..\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      "..\n",
      "..\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ". .\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ". .\n",
      ". .\n",
      ". .\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ". .\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      "....\n",
      ". .\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ". ..\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ". .\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".. .\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ". .\n",
      "..\n",
      ". .\n",
      ". .\n",
      ". .. .\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      "...\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      "...\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      "..\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      "..\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".. .\n",
      ".\n",
      ".\n",
      ".\n",
      ".. .\n",
      ". .\n",
      ".\n",
      ".\n",
      "..\n",
      ".\n",
      ".\n",
      ".\n",
      "..\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ". .\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      "⸮\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      "..\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ". ..\n",
      ".\n",
      ". .\n",
      ".\n",
      "...\n",
      ".\n",
      ".\n",
      ".\n",
      ". .\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      "..\n",
      ". .\n",
      ".\n",
      ".\n",
      ".\n",
      ". ..\n",
      ".\n",
      ".\n",
      ". .\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ". .\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ". .\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      "..\n",
      ".\n",
      ". .\n",
      ".\n",
      ".\n",
      "..\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ". ..\n",
      ".\n",
      ".\n",
      "..\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      "...\n",
      ".\n",
      ".\n",
      ". .\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      "...\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ". .\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      "...\n",
      ".\n",
      ". .\n",
      ".\n",
      ".\n",
      ".\n",
      "..\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      "..\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      "...\n",
      "..\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ". ..\n",
      ".\n",
      "..\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".. .\n",
      "⸮\n",
      ".\n",
      ".\n",
      "..\n",
      "...\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ". .. ..\n",
      ".\n",
      ". .\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".. .\n",
      "..\n",
      ".. .\n",
      ".\n",
      ".\n",
      ". .\n",
      ". ..\n",
      ".\n",
      ".\n",
      ".\n",
      ". .\n",
      ". .\n",
      ". .\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      "..\n",
      ". .\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".. .\n",
      ".\n",
      "..\n",
      "..\n",
      ".. ..\n",
      "...\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ". .\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ". . .\n",
      ".\n",
      ".\n",
      ".\n",
      "..\n",
      ".\n",
      ".\n",
      ". .\n",
      ".\n",
      "..\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ". .\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ". .\n",
      ".\n",
      ".\n",
      ". . ..\n",
      "..\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      ".\n",
      "\n",
      "iv Preface\n",
      "Human: hello\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "Hello! How can I assist you today?\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mGiven the following conversation and a follow up question, rephrase the follow up question to be a standalone question, in its original language.\n",
      "\n",
      "Chat History:\n",
      "\n",
      "Human: hello\n",
      "Assistant: Hello! How can I assist you today?\n",
      "Follow Up Input: what is genomic breeding\n",
      "Standalone question:\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "670 Theoretical and Applied Genetics (2019) 132:669–686\n",
      "1 3\n",
      "to select superior genotypes with a high agronomic perfor -\n",
      "mance across a range of different environmental conditions \n",
      "at the highest possible precision. This is a main driver of \n",
      "costs and time of the breeding cycle and ultimately limits the \n",
      "number of variety candidates that can be tested.\n",
      "A breeding technology that can potentially accelerate the \n",
      "rate of genetic gain in crops is genomic selection (GS) (Hef-\n",
      "fner et al. 2009; Meuwissen et al. 2001). To implement GS, \n",
      "a training population is first established, of individuals with \n",
      "phenotypes for the target trait(s) and genome-wide DNA \n",
      "marker genotypes (Meuwissen et al. 2001). The genotype \n",
      "and phenotype information from the training set is used to \n",
      "derive a prediction equation, which predicts the effect of \n",
      "each marker on the trait, with all marker effects fitted simul-\n",
      "taneously. If the markers are in sufficient linkage disequi -\n",
      "\n",
      "2021). They pinpoint the precise genetic differences con -\n",
      "nected to desirable qualities, making it possible to pick \n",
      "individuals with the best genomic profiles for breeding \n",
      "with accuracy and efficiency. Recent advances in genom -\n",
      "ics are producing new plant breeding methodologies and \n",
      "ways (e.g., association mapping, marker-assisted selec -\n",
      "tion, genomic selection, genome editing, etc.).\n",
      "Genomic resources for plant breeding encompass \n",
      "genetic markers, reference genomes, genomic and pro -\n",
      "tein databases, transcriptomes, and gene expression \n",
      "profiles. These resources facilitate the identification of \n",
      "genes associated with desirable traits, understanding \n",
      "genetic diversity, and acceleration of breeding programs \n",
      "[112]. Key techniques include genome-wide associa -\n",
      "tion studies (GWAS), marker-assisted selection (MAS), \n",
      "and genomic selection (GS), which allow for precise trait \n",
      "selection and prediction of breeding outcomes. By inte -\n",
      "\n",
      "Kumar et al. Biological Research           (2024) 57:80  \n",
      "https://doi.org/10.1186/s40659-024-00562-6\n",
      "REVIEW\n",
      "Advances in genomic tools for plant \n",
      "breeding: harnessing DNA molecular markers, \n",
      "genomic selection, and genome editing\n",
      "Rahul Kumar1*  , Sankar Prasad Das2, Burhan Uddin Choudhury1, Amit Kumar3, Nitish Ranjan Prakash4, \n",
      "Ramlakhan Verma5, Mridul Chakraborti5, Ayam Gangarani Devi1, Bijoya Bhattacharjee1, Rekha Das1, Bapi Das1, \n",
      "H. Lembisana Devi6, Biswajit Das1, Santoshi Rawat7 and Vinay Kumar Mishra3 \n",
      "Abstract \n",
      "Conventional pre-genomics breeding methodologies have significantly improved crop yields since the mid-twentieth \n",
      "century. Genomics provides breeders with advanced tools for whole-genome study, enabling a direct genotype–phe-\n",
      "notype analysis. This shift has led to precise and efficient crop development through genomics-based approaches, \n",
      "including molecular markers, genomic selection, and genome editing. Molecular markers, such as SNPs, are crucial\n",
      "\n",
      "Glossary\n",
      "Breeding value (BV): the worth of\n",
      "an individual as estimated by the\n",
      "average performance of its\n",
      "progenies.\n",
      "Array: high-throughput genotyping\n",
      "platform containing either genome-\n",
      "wide SNPs (SNP array) or coding\n",
      "variants (exome array).\n",
      "Genetic gain: the amount of\n",
      "increase in performance that is\n",
      "achieved through genetic\n",
      "improvement programs between\n",
      "consecutive selection cycles.\n",
      "Genomic estimated breeding\n",
      "value (GEBV): breeding value of an\n",
      "individual calculated using genome-\n",
      "wide marker data to capture small\n",
      "genetic effects dispersed over the\n",
      "genome.\n",
      "Genomic selection (GS): a type of\n",
      "breeding methodology that selects\n",
      "the best candidates as parents for\n",
      "the next selection cycle based on\n",
      "their predicted breeding values\n",
      "computed given: (i) their genotypes;\n",
      "and (ii) the phenotypes and\n",
      "genotypes of their relatives.\n",
      "Genotyping-by-sequencing (GBS):\n",
      "a highly multiplex genotyping system\n",
      "that involves DNA digestion with\n",
      "enzymes followed by construction of\n",
      "a reduced representation library,\n",
      "Human: Can you explain what genomic breeding is?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "Genomic breeding, also known as genomic selection (GS), is a modern plant\n",
      "breeding methodology that involves using molecular markers and genomic\n",
      "information to select the best candidates for breeding with high precision. By\n",
      "analyzing the genetic markers and whole-genome information of individuals,\n",
      "breeders can identify specific genes associated with desirable traits. This\n",
      "allows for accurate and efficient selection of individuals with the best genomic\n",
      "profiles for further breeding, ultimately accelerating the rate of genetic gain\n",
      "in crops and improving agronomic performance across different environmental\n",
      "conditions.\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mGiven the following conversation and a follow up question, rephrase the follow up question to be a standalone question, in its original language.\n",
      "\n",
      "Chat History:\n",
      "\n",
      "Human: hello\n",
      "Assistant: Hello! How can I assist you today?\n",
      "Human: what is genomic breeding\n",
      "Assistant: Genomic breeding, also known as genomic selection (GS), is a modern plant breeding methodology that involves using molecular markers and genomic information to select the best candidates for breeding with high precision. By analyzing the genetic markers and whole-genome information of individuals, breeders can identify specific genes associated with desirable traits. This allows for accurate and efficient selection of individuals with the best genomic profiles for further breeding, ultimately accelerating the rate of genetic gain in crops and improving agronomic performance across different environmental conditions.\n",
      "Follow Up Input: what is the disadvantages\n",
      "Standalone question:\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "process of GS is also explained inFigure 1.\n",
      "The major advantage of using GS is that it allows for a drastic\n",
      "reduction in the duration of the breeding cycle as compared to\n",
      "traditional breeding and also minimizes the cost associated with\n",
      "extensive phenotyping, thereby subsequently accelerating genetic\n",
      "gains and ensuring food and nutritional security (Heffner et al.,\n",
      "2010). However, there are certain factors such as the size of\n",
      "training and breeding populations, genetic diversity of breeding\n",
      "population, heritability of the underlying trait, inﬂuence of\n",
      "genotype– environment (GxE) interaction, density of markers,\n",
      "and genetic relationship between training population and\n",
      "breeding population or selection candidates, which may\n",
      "inﬂuence the genomic prediction’s accuracy (De Roos et al.,\n",
      "2009; Lorenzana and Bernardo, 2009 ; Luan et al., 2009 ;\n",
      "Daetwyler et al., 2010; Clark et al., 2011; Howard et al., 2014).\n",
      "Hence, successful implementation of GS in breeding programs\n",
      "\n",
      "680 Theoretical and Applied Genetics (2019) 132:669–686\n",
      "1 3\n",
      "associated with DHs are that the technology can be costly, \n",
      "at times inefficient in terms of natural or chemical-induced \n",
      "chromosome doubling and the breeder cannot select for \n",
      "basic traits during line development. Rapid generation \n",
      "advance techniques like “speed breeding” (Watson et al. \n",
      "2018) are gaining popularity because they allow the breeder \n",
      "to turnover many generations under glasshouse conditions \n",
      "while enabling the breeder to select for traits with a high \n",
      "heritability. By modifying temperature and light regimes, \n",
      "plant growth can be greatly accelerated. For wheat, barley, \n",
      "or rapeseed for example, this enables the breeder to develop \n",
      "F6 lines, suitable for yield field trials, in only 1–1.5 years, \n",
      "thereby significantly reducing the line development phase of \n",
      "a breeding programme. While under speed breeding condi-\n",
      "tions breeders can discard variety candidates with unfavour-\n",
      "\n",
      "670 Theoretical and Applied Genetics (2019) 132:669–686\n",
      "1 3\n",
      "to select superior genotypes with a high agronomic perfor -\n",
      "mance across a range of different environmental conditions \n",
      "at the highest possible precision. This is a main driver of \n",
      "costs and time of the breeding cycle and ultimately limits the \n",
      "number of variety candidates that can be tested.\n",
      "A breeding technology that can potentially accelerate the \n",
      "rate of genetic gain in crops is genomic selection (GS) (Hef-\n",
      "fner et al. 2009; Meuwissen et al. 2001). To implement GS, \n",
      "a training population is first established, of individuals with \n",
      "phenotypes for the target trait(s) and genome-wide DNA \n",
      "marker genotypes (Meuwissen et al. 2001). The genotype \n",
      "and phenotype information from the training set is used to \n",
      "derive a prediction equation, which predicts the effect of \n",
      "each marker on the trait, with all marker effects fitted simul-\n",
      "taneously. If the markers are in sufficient linkage disequi -\n",
      "\n",
      "Kumar et al. Biological Research           (2024) 57:80  \n",
      "https://doi.org/10.1186/s40659-024-00562-6\n",
      "REVIEW\n",
      "Advances in genomic tools for plant \n",
      "breeding: harnessing DNA molecular markers, \n",
      "genomic selection, and genome editing\n",
      "Rahul Kumar1*  , Sankar Prasad Das2, Burhan Uddin Choudhury1, Amit Kumar3, Nitish Ranjan Prakash4, \n",
      "Ramlakhan Verma5, Mridul Chakraborti5, Ayam Gangarani Devi1, Bijoya Bhattacharjee1, Rekha Das1, Bapi Das1, \n",
      "H. Lembisana Devi6, Biswajit Das1, Santoshi Rawat7 and Vinay Kumar Mishra3 \n",
      "Abstract \n",
      "Conventional pre-genomics breeding methodologies have significantly improved crop yields since the mid-twentieth \n",
      "century. Genomics provides breeders with advanced tools for whole-genome study, enabling a direct genotype–phe-\n",
      "notype analysis. This shift has led to precise and efficient crop development through genomics-based approaches, \n",
      "including molecular markers, genomic selection, and genome editing. Molecular markers, such as SNPs, are crucial\n",
      "Human: What are the disadvantages of genomic breeding?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "Some of the disadvantages of genomic selection in breeding programs include\n",
      "factors like the cost of technology, potential inefficiency in chromosome\n",
      "doubling, limitations in selecting for basic traits during line development, and\n",
      "challenges in predicting accuracy influenced by various factors such as training\n",
      "and breeding population size, genetic diversity, genotype-environment\n",
      "interactions, density of markers, and genetic relationships between populations.\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mGiven the following conversation and a follow up question, rephrase the follow up question to be a standalone question, in its original language.\n",
      "\n",
      "Chat History:\n",
      "\n",
      "Human: hello\n",
      "Assistant: Hello! How can I assist you today?\n",
      "Human: what is genomic breeding\n",
      "Assistant: Genomic breeding, also known as genomic selection (GS), is a modern plant breeding methodology that involves using molecular markers and genomic information to select the best candidates for breeding with high precision. By analyzing the genetic markers and whole-genome information of individuals, breeders can identify specific genes associated with desirable traits. This allows for accurate and efficient selection of individuals with the best genomic profiles for further breeding, ultimately accelerating the rate of genetic gain in crops and improving agronomic performance across different environmental conditions.\n",
      "Human: what is the disadvantages\n",
      "Assistant: Some of the disadvantages of genomic selection in breeding programs include factors like the cost of technology, potential inefficiency in chromosome doubling, limitations in selecting for basic traits during line development, and challenges in predicting accuracy influenced by various factors such as training and breeding population size, genetic diversity, genotype-environment interactions, density of markers, and genetic relationships between populations.\n",
      "Follow Up Input: advantages?\n",
      "Standalone question:\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "2021). They pinpoint the precise genetic differences con -\n",
      "nected to desirable qualities, making it possible to pick \n",
      "individuals with the best genomic profiles for breeding \n",
      "with accuracy and efficiency. Recent advances in genom -\n",
      "ics are producing new plant breeding methodologies and \n",
      "ways (e.g., association mapping, marker-assisted selec -\n",
      "tion, genomic selection, genome editing, etc.).\n",
      "Genomic resources for plant breeding encompass \n",
      "genetic markers, reference genomes, genomic and pro -\n",
      "tein databases, transcriptomes, and gene expression \n",
      "profiles. These resources facilitate the identification of \n",
      "genes associated with desirable traits, understanding \n",
      "genetic diversity, and acceleration of breeding programs \n",
      "[112]. Key techniques include genome-wide associa -\n",
      "tion studies (GWAS), marker-assisted selection (MAS), \n",
      "and genomic selection (GS), which allow for precise trait \n",
      "selection and prediction of breeding outcomes. By inte -\n",
      "\n",
      "crop improvement.Genomics approaches are beneficial \n",
      "when dealing with complex traits, as these traits usu -\n",
      "ally have a multi-genic nature and a significant environ -\n",
      "mental influence [102]. Genomic tools provide genomic \n",
      "information and facilitate the detection of QTLs and the \n",
      "identification of existing favorable alleles of small effect, \n",
      "which have frequently remained unnoticed and have not \n",
      "been included in the gene pool used for breeding [102]. \n",
      "Genomic tools have revolutionized plant breeding by \n",
      "enabling more precise, efficient, and targeted approaches \n",
      "to developing new plant varieties through genome-wide \n",
      "association study, marker-assisted selection, genomic \n",
      "selection, and gene editing. In this review, we present and \n",
      "discuss the most relevant advances in the development \n",
      "of genomic tools and provide examples of applying these \n",
      "tools to plant breeding.\n",
      "Molecular markers: tool for the genetic analysis\n",
      "Based on nucleotide sequence polymorphisms, molecu -\n",
      "\n",
      "oping improved crop cultivars. However, the expeditious emer-\n",
      "gence of DNA-sequencing technology has allowed breeders to\n",
      "gain comprehensive genomic information on crops, which is\n",
      "very valuable for selection. The development of several DNA-\n",
      "marker-based genotyping systems signiﬁcantly increased the\n",
      "number of DNA markers available to plant breeders (Crossa\n",
      "et al., 2017). This breakthrough allowed plant breeders to select\n",
      "plant performance based on their genetic marker composition\n",
      "rather than solely on their phenotypic performance, which is\n",
      "prone to several limitations in selection efﬁciency.\n",
      "The application of genomic tools in the breeding practice of\n",
      "plants, generally termed genomic-assisted breeding, has pro-\n",
      "gressed through various stages in the last four decades\n",
      "(Varshney et al., 2021). It started with linkage-based mapping\n",
      "of quantitative trait loci (QTLs) (Soller and Plotkin-Hazan, 1977)\n",
      "where, with a limited number of DNA markers, those\n",
      "\n",
      "selection and prediction of breeding outcomes. By inte -\n",
      "grating these genomic tools, plant breeders can improve \n",
      "crop yield, disease resistance, and stress tolerance, \n",
      "enhancing agricultural productivity and sustainability. \n",
      "The existing variability among crop species is utilized \n",
      "for plant breeding activities, which can also be generated \n",
      "through crossing or induced mutagenesis. In addition to \n",
      "the identification of genetic markers and the availability \n",
      "of published genomes, clustered regularly interspaced \n",
      "short palindromic repeats-associated protein 9 (CRISPR/\n",
      "Cas9) is promising for application to modern breeding \n",
      "and is a novel technology for genome editing in major \n",
      "crops [132]. CRISPR/Cas9-based directional breeding is \n",
      "highly efficient and saves more time than other breeding \n",
      "techniques that use genome editing. It further enables \n",
      "targeted genetic modifications, opening new avenues for \n",
      "crop improvement.Genomics approaches are beneficial\n",
      "Human: What are some of the advantages of genomic breeding in plant breeding programs?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "Some advantages of genomic breeding in plant breeding programs include:  1.\n",
      "**Precision**: Genomic tools allow for precise selection of desirable traits and\n",
      "prediction of breeding outcomes.     2. **Efficiency**: Genomic breeding methods\n",
      "enable more efficient selection and breeding processes, saving time and\n",
      "resources.     3. **Targeted Approach**: Genomic tools facilitate targeted\n",
      "genetic modifications, improving the accuracy of breeding programs.     4.\n",
      "**Identifying Favorable Alleles**: Genomic tools help in identifying existing\n",
      "favorable alleles of small effect that may have been overlooked previously.\n",
      "5. **Accelerating Breeding Programs**: Genomic resources accelerate breeding\n",
      "programs by enabling the identification of genes associated with desirable\n",
      "traits.     6. **Enhancing Agricultural Productivity**: By improving crop yield,\n",
      "disease resistance, and stress tolerance, genomic breeding enhances agricultural\n",
      "productivity and sustainability.  These advantages ultimately contribute to the\n",
      "development of new plant varieties with improved traits in a more precise,\n",
      "efficient, and targeted manner.\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mGiven the following conversation and a follow up question, rephrase the follow up question to be a standalone question, in its original language.\n",
      "\n",
      "Chat History:\n",
      "\n",
      "Human: hello\n",
      "Assistant: Hello! How can I assist you today?\n",
      "Human: what is genomic breeding\n",
      "Assistant: Genomic breeding, also known as genomic selection (GS), is a modern plant breeding methodology that involves using molecular markers and genomic information to select the best candidates for breeding with high precision. By analyzing the genetic markers and whole-genome information of individuals, breeders can identify specific genes associated with desirable traits. This allows for accurate and efficient selection of individuals with the best genomic profiles for further breeding, ultimately accelerating the rate of genetic gain in crops and improving agronomic performance across different environmental conditions.\n",
      "Human: what is the disadvantages\n",
      "Assistant: Some of the disadvantages of genomic selection in breeding programs include factors like the cost of technology, potential inefficiency in chromosome doubling, limitations in selecting for basic traits during line development, and challenges in predicting accuracy influenced by various factors such as training and breeding population size, genetic diversity, genotype-environment interactions, density of markers, and genetic relationships between populations.\n",
      "Human: advantages?\n",
      "Assistant: Some advantages of genomic breeding in plant breeding programs include:\n",
      "\n",
      "1. **Precision**: Genomic tools allow for precise selection of desirable traits and prediction of breeding outcomes.\n",
      "   \n",
      "2. **Efficiency**: Genomic breeding methods enable more efficient selection and breeding processes, saving time and resources.\n",
      "   \n",
      "3. **Targeted Approach**: Genomic tools facilitate targeted genetic modifications, improving the accuracy of breeding programs.\n",
      "   \n",
      "4. **Identifying Favorable Alleles**: Genomic tools help in identifying existing favorable alleles of small effect that may have been overlooked previously.\n",
      "   \n",
      "5. **Accelerating Breeding Programs**: Genomic resources accelerate breeding programs by enabling the identification of genes associated with desirable traits.\n",
      "   \n",
      "6. **Enhancing Agricultural Productivity**: By improving crop yield, disease resistance, and stress tolerance, genomic breeding enhances agricultural productivity and sustainability.\n",
      "\n",
      "These advantages ultimately contribute to the development of new plant varieties with improved traits in a more precise, efficient, and targeted manner.\n",
      "Follow Up Input: what should we pay attentation when doing that\n",
      "Standalone question:\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "Genomic selection in plant breeding: Key factors\n",
      "shaping two decades of progress\n",
      "Admas Alemu1,*, Johanna A˚strand1,2, Osval A. Montesinos-Lo´ pez3, Julio Isidro y Sa´ nchez4,\n",
      "Javier Ferna´ ndez-Go´ nzalez4, Wuletaw Tadesse5, Ramesh R. Vetukuri1, Anders S. Carlsson1,\n",
      "Alf Ceplitis2, Jose´ Crossa6, Rodomiro Ortiz1,* and Aakash Chawade1\n",
      "1Department of Plant Breeding, Swedish University of Agricultural Sciences, Alnarp, Sweden\n",
      "2Lantm€annen Lantbruk, Sval€ov, Sweden\n",
      "3Facultad de Telema´ tica, University de Colima, Colima, Colima 28040, Mexico\n",
      "4Centro de Biotecnologı´ a y Geno´ mica de Plantas (CBGP, UPM-INIA), Universidad Polite´ cnica de Madrid (UPM) – Instituto Nacional de Investigacio´ n y Tecnologı´ a\n",
      "Agraria y Alimentaria (INIA), Campus de Montegancedo-UPM, 28223 Madrid, Spain\n",
      "5International Center for Agricultural Research in the Dry Areas (ICARDA), Rabat, Morocco\n",
      "\n",
      "plant breeding could be limited by some of the following factors: (i) genotyping cost,\n",
      "(ii) unclear guidelines about where in the breeding program GS could be ef ﬁciently\n",
      "applied (Crossa et al. 2017), (iii) insuf ﬁcient number of lines or animals in the\n",
      "reference (training) population, (iv) insuf ﬁcient number of SNPs in the panels, and\n",
      "(v) the reference population contains very heterogeneous individuals (plants or\n",
      "animals).\n",
      "1.2.2 Why Is Statistical Machine Learning a Key Element\n",
      "of Genomic Selection?\n",
      "GS is challenging and very interesting because it aims to improve crop productivity\n",
      "to satisfy humankind ’s need for food. Addressing the current challenges to increase\n",
      "crop productivity by improving the genetic makeup of plants and avoiding plant\n",
      "diseases is not new, but it is of paramount importance today to be able to increase\n",
      "crop productivity around the world without the need to increase the arable land.\n",
      "Statistical machine learning methods can help improve GS methodology, since they\n",
      "\n",
      "2021). They pinpoint the precise genetic differences con -\n",
      "nected to desirable qualities, making it possible to pick \n",
      "individuals with the best genomic profiles for breeding \n",
      "with accuracy and efficiency. Recent advances in genom -\n",
      "ics are producing new plant breeding methodologies and \n",
      "ways (e.g., association mapping, marker-assisted selec -\n",
      "tion, genomic selection, genome editing, etc.).\n",
      "Genomic resources for plant breeding encompass \n",
      "genetic markers, reference genomes, genomic and pro -\n",
      "tein databases, transcriptomes, and gene expression \n",
      "profiles. These resources facilitate the identification of \n",
      "genes associated with desirable traits, understanding \n",
      "genetic diversity, and acceleration of breeding programs \n",
      "[112]. Key techniques include genome-wide associa -\n",
      "tion studies (GWAS), marker-assisted selection (MAS), \n",
      "and genomic selection (GS), which allow for precise trait \n",
      "selection and prediction of breeding outcomes. By inte -\n",
      "\n",
      "crop improvement.Genomics approaches are beneficial \n",
      "when dealing with complex traits, as these traits usu -\n",
      "ally have a multi-genic nature and a significant environ -\n",
      "mental influence [102]. Genomic tools provide genomic \n",
      "information and facilitate the detection of QTLs and the \n",
      "identification of existing favorable alleles of small effect, \n",
      "which have frequently remained unnoticed and have not \n",
      "been included in the gene pool used for breeding [102]. \n",
      "Genomic tools have revolutionized plant breeding by \n",
      "enabling more precise, efficient, and targeted approaches \n",
      "to developing new plant varieties through genome-wide \n",
      "association study, marker-assisted selection, genomic \n",
      "selection, and gene editing. In this review, we present and \n",
      "discuss the most relevant advances in the development \n",
      "of genomic tools and provide examples of applying these \n",
      "tools to plant breeding.\n",
      "Molecular markers: tool for the genetic analysis\n",
      "Based on nucleotide sequence polymorphisms, molecu -\n",
      "Human: What factors should we pay attention to when implementing genomic breeding in plant breeding programs?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "When implementing genomic breeding in plant breeding programs, it is important\n",
      "to pay attention to the following factors:  1. Genotyping cost 2. Clear\n",
      "guidelines on where in the breeding program genomic selection (GS) could be\n",
      "efficiently applied 3. Sufficient number of lines or animals in the reference\n",
      "(training) population 4. Sufficient number of SNPs in the panels 5. The\n",
      "homogeneity of individuals within the reference population (plants or animals)\n",
      "These factors are key to ensuring the success and efficiency of genomic\n",
      "selection in plant breeding programs.\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mGiven the following conversation and a follow up question, rephrase the follow up question to be a standalone question, in its original language.\n",
      "\n",
      "Chat History:\n",
      "\n",
      "Human: hello\n",
      "Assistant: Hello! How can I assist you today?\n",
      "Human: what is genomic breeding\n",
      "Assistant: Genomic breeding, also known as genomic selection (GS), is a modern plant breeding methodology that involves using molecular markers and genomic information to select the best candidates for breeding with high precision. By analyzing the genetic markers and whole-genome information of individuals, breeders can identify specific genes associated with desirable traits. This allows for accurate and efficient selection of individuals with the best genomic profiles for further breeding, ultimately accelerating the rate of genetic gain in crops and improving agronomic performance across different environmental conditions.\n",
      "Human: what is the disadvantages\n",
      "Assistant: Some of the disadvantages of genomic selection in breeding programs include factors like the cost of technology, potential inefficiency in chromosome doubling, limitations in selecting for basic traits during line development, and challenges in predicting accuracy influenced by various factors such as training and breeding population size, genetic diversity, genotype-environment interactions, density of markers, and genetic relationships between populations.\n",
      "Human: advantages?\n",
      "Assistant: Some advantages of genomic breeding in plant breeding programs include:\n",
      "\n",
      "1. **Precision**: Genomic tools allow for precise selection of desirable traits and prediction of breeding outcomes.\n",
      "   \n",
      "2. **Efficiency**: Genomic breeding methods enable more efficient selection and breeding processes, saving time and resources.\n",
      "   \n",
      "3. **Targeted Approach**: Genomic tools facilitate targeted genetic modifications, improving the accuracy of breeding programs.\n",
      "   \n",
      "4. **Identifying Favorable Alleles**: Genomic tools help in identifying existing favorable alleles of small effect that may have been overlooked previously.\n",
      "   \n",
      "5. **Accelerating Breeding Programs**: Genomic resources accelerate breeding programs by enabling the identification of genes associated with desirable traits.\n",
      "   \n",
      "6. **Enhancing Agricultural Productivity**: By improving crop yield, disease resistance, and stress tolerance, genomic breeding enhances agricultural productivity and sustainability.\n",
      "\n",
      "These advantages ultimately contribute to the development of new plant varieties with improved traits in a more precise, efficient, and targeted manner.\n",
      "Human: what should we pay attentation when doing that\n",
      "Assistant: When implementing genomic breeding in plant breeding programs, it is important to pay attention to the following factors:\n",
      "\n",
      "1. Genotyping cost\n",
      "2. Clear guidelines on where in the breeding program genomic selection (GS) could be efficiently applied\n",
      "3. Sufficient number of lines or animals in the reference (training) population\n",
      "4. Sufficient number of SNPs in the panels\n",
      "5. The homogeneity of individuals within the reference population (plants or animals)\n",
      "\n",
      "These factors are key to ensuring the success and efficiency of genomic selection in plant breeding programs.\n",
      "Follow Up Input: what is the machine learning in GS\n",
      "Standalone question:\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "plant breeding could be limited by some of the following factors: (i) genotyping cost,\n",
      "(ii) unclear guidelines about where in the breeding program GS could be ef ﬁciently\n",
      "applied (Crossa et al. 2017), (iii) insuf ﬁcient number of lines or animals in the\n",
      "reference (training) population, (iv) insuf ﬁcient number of SNPs in the panels, and\n",
      "(v) the reference population contains very heterogeneous individuals (plants or\n",
      "animals).\n",
      "1.2.2 Why Is Statistical Machine Learning a Key Element\n",
      "of Genomic Selection?\n",
      "GS is challenging and very interesting because it aims to improve crop productivity\n",
      "to satisfy humankind ’s need for food. Addressing the current challenges to increase\n",
      "crop productivity by improving the genetic makeup of plants and avoiding plant\n",
      "diseases is not new, but it is of paramount importance today to be able to increase\n",
      "crop productivity around the world without the need to increase the arable land.\n",
      "Statistical machine learning methods can help improve GS methodology, since they\n",
      "\n",
      "have been created to evaluate genomic prediction accu -\n",
      "racy and make genomic selection (GS) usable. Some nota-\n",
      "ble tools and approaches are rrBLUP , solGS, GMStool, \n",
      "BWGS, BGLR, GenSel, GSelection and lme4GS [18]. \n",
      "Genomic Selection (GS) revolutionizes crop improve -\n",
      "ment, streamlining breeding for precision and effi -\n",
      "ciency. Utilizing genetic markers, GS predicts crop yields \n",
      "(Table 4). For example, Cerioli et al. [21] used the LSU500 \n",
      "marker set to achieve predictive abilities of 0.13 to 0.78 \n",
      "in various rice trials. GS is equally potent in disease \n",
      "resistance breeding, as seen in Zhang et al. ’s [189] study \n",
      "on Fusarium head blight (FHB) traits in wheat, yielding \n",
      "prediction accuracies of up to 0.59 for FHB disease index \n",
      "and 0.54 for Fusarium damaged Kernels (FDK). Further -\n",
      "more, GS enhances abiotic stress tolerance. Zhang et al. \n",
      "[190] assessed maize drought tolerance with prediction \n",
      "accuracies ranging from 0.19 to 0.33 for traits like seed -\n",
      "\n",
      "Preface\n",
      "In plant breeding, prediction models are essential for the successful implementation\n",
      "of genomic selection (GS), which is considered a predictive methodology used to\n",
      "train models with a reference training population containing known phenotypic and\n",
      "genotypic data and then perform predictions for a testing data set that only contains\n",
      "genomic data. However, because a universal model is nonexistent (free lunch\n",
      "theorem), it is necessary to evaluate many models for a particular data set and\n",
      "subsequently choose the best option for each particular situation. Thus, a great\n",
      "variety of statistical models are used for prediction in GS. The GS genome-based\n",
      "statistical machine prediction models face several complexities, as the choice of the\n",
      "model depends on multiple factors such as the genetic architecture of the trait,\n",
      "marker density, sample size, the span of linkage disequilibrium, and the geno-\n",
      "type /C2environment interaction.\n",
      "\n",
      "genotypic) needs to be improved. Many times the genotypic information available is\n",
      "not enough to make high-quality predictions of the target trait, since the information\n",
      "available has a lot of noise. Also, since there is no universal best prediction model\n",
      "that can be used under all circumstances, a good understanding of statistical machine\n",
      "learning models is required to increase the ef ﬁciency of the selection process of the\n",
      "best candidate individuals with GS early in time. This is very important because one\n",
      "of the key components of genomic selection is the use of statistical machine learning\n",
      "models for the prediction of non-phenotyped individuals. For this reason, statistical\n",
      "machine learning tools have the power to help increase the potential of GS if more\n",
      "powerful statistical machine learning methods are developed, if the existing methods\n",
      "can deal with larger data sets, and if these methods can be automatized to perform the\n",
      "Human: What role does machine learning play in genomic selection (GS)?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "Machine learning plays a key role in genomic selection (GS) by helping to\n",
      "improve the methodology for genomic prediction accuracy and making GS usable.\n",
      "Machine learning methods are essential for evaluating genomic prediction\n",
      "accuracy, training models with reference populations, and predicting traits in\n",
      "non-phenotyped individuals. These methods are crucial in selecting the best\n",
      "candidates early in the breeding process, especially when genotypic information\n",
      "is limited or noisy. Therefore, machine learning tools contribute to enhancing\n",
      "the potential of GS by enabling more efficient and precise selection of\n",
      "candidate individuals.\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mGiven the following conversation and a follow up question, rephrase the follow up question to be a standalone question, in its original language.\n",
      "\n",
      "Chat History:\n",
      "\n",
      "Human: hello\n",
      "Assistant: Hello! How can I assist you today?\n",
      "Human: what is genomic breeding\n",
      "Assistant: Genomic breeding, also known as genomic selection (GS), is a modern plant breeding methodology that involves using molecular markers and genomic information to select the best candidates for breeding with high precision. By analyzing the genetic markers and whole-genome information of individuals, breeders can identify specific genes associated with desirable traits. This allows for accurate and efficient selection of individuals with the best genomic profiles for further breeding, ultimately accelerating the rate of genetic gain in crops and improving agronomic performance across different environmental conditions.\n",
      "Human: what is the disadvantages\n",
      "Assistant: Some of the disadvantages of genomic selection in breeding programs include factors like the cost of technology, potential inefficiency in chromosome doubling, limitations in selecting for basic traits during line development, and challenges in predicting accuracy influenced by various factors such as training and breeding population size, genetic diversity, genotype-environment interactions, density of markers, and genetic relationships between populations.\n",
      "Human: advantages?\n",
      "Assistant: Some advantages of genomic breeding in plant breeding programs include:\n",
      "\n",
      "1. **Precision**: Genomic tools allow for precise selection of desirable traits and prediction of breeding outcomes.\n",
      "   \n",
      "2. **Efficiency**: Genomic breeding methods enable more efficient selection and breeding processes, saving time and resources.\n",
      "   \n",
      "3. **Targeted Approach**: Genomic tools facilitate targeted genetic modifications, improving the accuracy of breeding programs.\n",
      "   \n",
      "4. **Identifying Favorable Alleles**: Genomic tools help in identifying existing favorable alleles of small effect that may have been overlooked previously.\n",
      "   \n",
      "5. **Accelerating Breeding Programs**: Genomic resources accelerate breeding programs by enabling the identification of genes associated with desirable traits.\n",
      "   \n",
      "6. **Enhancing Agricultural Productivity**: By improving crop yield, disease resistance, and stress tolerance, genomic breeding enhances agricultural productivity and sustainability.\n",
      "\n",
      "These advantages ultimately contribute to the development of new plant varieties with improved traits in a more precise, efficient, and targeted manner.\n",
      "Human: what should we pay attentation when doing that\n",
      "Assistant: When implementing genomic breeding in plant breeding programs, it is important to pay attention to the following factors:\n",
      "\n",
      "1. Genotyping cost\n",
      "2. Clear guidelines on where in the breeding program genomic selection (GS) could be efficiently applied\n",
      "3. Sufficient number of lines or animals in the reference (training) population\n",
      "4. Sufficient number of SNPs in the panels\n",
      "5. The homogeneity of individuals within the reference population (plants or animals)\n",
      "\n",
      "These factors are key to ensuring the success and efficiency of genomic selection in plant breeding programs.\n",
      "Human: what is the machine learning in GS\n",
      "Assistant: Machine learning plays a key role in genomic selection (GS) by helping to improve the methodology for genomic prediction accuracy and making GS usable. Machine learning methods are essential for evaluating genomic prediction accuracy, training models with reference populations, and predicting traits in non-phenotyped individuals. These methods are crucial in selecting the best candidates early in the breeding process, especially when genotypic information is limited or noisy. Therefore, machine learning tools contribute to enhancing the potential of GS by enabling more efficient and precise selection of candidate individuals.\n",
      "Follow Up Input: give examples and references\n",
      "Standalone question:\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "plant breeding could be limited by some of the following factors: (i) genotyping cost,\n",
      "(ii) unclear guidelines about where in the breeding program GS could be ef ﬁciently\n",
      "applied (Crossa et al. 2017), (iii) insuf ﬁcient number of lines or animals in the\n",
      "reference (training) population, (iv) insuf ﬁcient number of SNPs in the panels, and\n",
      "(v) the reference population contains very heterogeneous individuals (plants or\n",
      "animals).\n",
      "1.2.2 Why Is Statistical Machine Learning a Key Element\n",
      "of Genomic Selection?\n",
      "GS is challenging and very interesting because it aims to improve crop productivity\n",
      "to satisfy humankind ’s need for food. Addressing the current challenges to increase\n",
      "crop productivity by improving the genetic makeup of plants and avoiding plant\n",
      "diseases is not new, but it is of paramount importance today to be able to increase\n",
      "crop productivity around the world without the need to increase the arable land.\n",
      "Statistical machine learning methods can help improve GS methodology, since they\n",
      "\n",
      "have been created to evaluate genomic prediction accu -\n",
      "racy and make genomic selection (GS) usable. Some nota-\n",
      "ble tools and approaches are rrBLUP , solGS, GMStool, \n",
      "BWGS, BGLR, GenSel, GSelection and lme4GS [18]. \n",
      "Genomic Selection (GS) revolutionizes crop improve -\n",
      "ment, streamlining breeding for precision and effi -\n",
      "ciency. Utilizing genetic markers, GS predicts crop yields \n",
      "(Table 4). For example, Cerioli et al. [21] used the LSU500 \n",
      "marker set to achieve predictive abilities of 0.13 to 0.78 \n",
      "in various rice trials. GS is equally potent in disease \n",
      "resistance breeding, as seen in Zhang et al. ’s [189] study \n",
      "on Fusarium head blight (FHB) traits in wheat, yielding \n",
      "prediction accuracies of up to 0.59 for FHB disease index \n",
      "and 0.54 for Fusarium damaged Kernels (FDK). Further -\n",
      "more, GS enhances abiotic stress tolerance. Zhang et al. \n",
      "[190] assessed maize drought tolerance with prediction \n",
      "accuracies ranging from 0.19 to 0.33 for traits like seed -\n",
      "\n",
      "Preface\n",
      "In plant breeding, prediction models are essential for the successful implementation\n",
      "of genomic selection (GS), which is considered a predictive methodology used to\n",
      "train models with a reference training population containing known phenotypic and\n",
      "genotypic data and then perform predictions for a testing data set that only contains\n",
      "genomic data. However, because a universal model is nonexistent (free lunch\n",
      "theorem), it is necessary to evaluate many models for a particular data set and\n",
      "subsequently choose the best option for each particular situation. Thus, a great\n",
      "variety of statistical models are used for prediction in GS. The GS genome-based\n",
      "statistical machine prediction models face several complexities, as the choice of the\n",
      "model depends on multiple factors such as the genetic architecture of the trait,\n",
      "marker density, sample size, the span of linkage disequilibrium, and the geno-\n",
      "type /C2environment interaction.\n",
      "\n",
      "Feature Review\n",
      "Genomic Selection in Plant\n",
      "Breeding: Methods, Models,\n",
      "and Perspectives\n",
      "José Crossa,1,* Paulino Pérez-Rodríguez,2 Jaime Cuevas,3\n",
      "Osval Montesinos-López,4 Diego Jarquín,5\n",
      "Gustavo de los Campos,6 Juan Burgueño,1\n",
      "Juan M. González-Camacho,2 Sergio Pérez-Elizalde,2\n",
      "Yoseph Beyene,1 Susanne Dreisigacker,1 Ravi Singh,1\n",
      "Xuecai Zhang,1 Manje Gowda,1 Manish Roorkiwal,7\n",
      "Jessica Rutkoski,8 and Rajeev K. Varshney7,*\n",
      "Genomic selection (GS) facilitates the rapid selection of superior genotypes\n",
      "and accelerates the breeding cycle. In this review, we discuss the history,\n",
      "principles, and basis of GS and genomic-enabled prediction (GP) as well as the\n",
      "genetics and statistical complexities of GP models, including genomic geno-\n",
      "type /C2  environment (G /C2  E) interactions. We also examine the accuracy of GP\n",
      "models and methods for two cereal crops and two legume crops based on\n",
      "random cross-validation. GS applied to maize breeding has shown tangible\n",
      "Human: Can you provide examples and references for the role of machine learning in genomic selection (GS)?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "Sure! Some notable tools and approaches mentioned in the provided context that\n",
      "utilize machine learning in genomic selection (GS) are: 1. rrBLUP 2. solGS 3.\n",
      "GMStool 4. BWGS 5. BGLR 6. GenSel 7. GSelection 8. lme4GS  These tools and\n",
      "approaches are used to evaluate genomic prediction accuracy and make genomic\n",
      "selection usable. For example, Cerioli et al. used the LSU500 marker set to\n",
      "achieve predictive abilities in various rice trials, and Zhang et al. studied\n",
      "Fusarium head blight (FHB) traits in wheat using GS. Zhang et al. also assessed\n",
      "maize drought tolerance using GS.  Reference: - José Crossa, Paulino Pérez-\n",
      "Rodríguez, Jaime Cuevas, Osval Montesinos-López, Diego Jarquín, Gustavo de los\n",
      "Campos, Juan Burgueño, Juan M. González-Camacho, Sergio Pérez-Elizalde, Yoseph\n",
      "Beyene, Susanne Dreisigacker, Ravi Singh, Xuecai Zhang, Manje Gowda, Manish\n",
      "Roorkiwal, Jessica Rutkoski, and Rajeev K. Varshney. \"Genomic Selection in Plant\n",
      "Breeding: Methods, Models, and Perspectives.\"\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mGiven the following conversation and a follow up question, rephrase the follow up question to be a standalone question, in its original language.\n",
      "\n",
      "Chat History:\n",
      "\n",
      "Human: hello\n",
      "Assistant: Hello! How can I assist you today?\n",
      "Human: what is genomic breeding\n",
      "Assistant: Genomic breeding, also known as genomic selection (GS), is a modern plant breeding methodology that involves using molecular markers and genomic information to select the best candidates for breeding with high precision. By analyzing the genetic markers and whole-genome information of individuals, breeders can identify specific genes associated with desirable traits. This allows for accurate and efficient selection of individuals with the best genomic profiles for further breeding, ultimately accelerating the rate of genetic gain in crops and improving agronomic performance across different environmental conditions.\n",
      "Human: what is the disadvantages\n",
      "Assistant: Some of the disadvantages of genomic selection in breeding programs include factors like the cost of technology, potential inefficiency in chromosome doubling, limitations in selecting for basic traits during line development, and challenges in predicting accuracy influenced by various factors such as training and breeding population size, genetic diversity, genotype-environment interactions, density of markers, and genetic relationships between populations.\n",
      "Human: advantages?\n",
      "Assistant: Some advantages of genomic breeding in plant breeding programs include:\n",
      "\n",
      "1. **Precision**: Genomic tools allow for precise selection of desirable traits and prediction of breeding outcomes.\n",
      "   \n",
      "2. **Efficiency**: Genomic breeding methods enable more efficient selection and breeding processes, saving time and resources.\n",
      "   \n",
      "3. **Targeted Approach**: Genomic tools facilitate targeted genetic modifications, improving the accuracy of breeding programs.\n",
      "   \n",
      "4. **Identifying Favorable Alleles**: Genomic tools help in identifying existing favorable alleles of small effect that may have been overlooked previously.\n",
      "   \n",
      "5. **Accelerating Breeding Programs**: Genomic resources accelerate breeding programs by enabling the identification of genes associated with desirable traits.\n",
      "   \n",
      "6. **Enhancing Agricultural Productivity**: By improving crop yield, disease resistance, and stress tolerance, genomic breeding enhances agricultural productivity and sustainability.\n",
      "\n",
      "These advantages ultimately contribute to the development of new plant varieties with improved traits in a more precise, efficient, and targeted manner.\n",
      "Human: what should we pay attentation when doing that\n",
      "Assistant: When implementing genomic breeding in plant breeding programs, it is important to pay attention to the following factors:\n",
      "\n",
      "1. Genotyping cost\n",
      "2. Clear guidelines on where in the breeding program genomic selection (GS) could be efficiently applied\n",
      "3. Sufficient number of lines or animals in the reference (training) population\n",
      "4. Sufficient number of SNPs in the panels\n",
      "5. The homogeneity of individuals within the reference population (plants or animals)\n",
      "\n",
      "These factors are key to ensuring the success and efficiency of genomic selection in plant breeding programs.\n",
      "Human: what is the machine learning in GS\n",
      "Assistant: Machine learning plays a key role in genomic selection (GS) by helping to improve the methodology for genomic prediction accuracy and making GS usable. Machine learning methods are essential for evaluating genomic prediction accuracy, training models with reference populations, and predicting traits in non-phenotyped individuals. These methods are crucial in selecting the best candidates early in the breeding process, especially when genotypic information is limited or noisy. Therefore, machine learning tools contribute to enhancing the potential of GS by enabling more efficient and precise selection of candidate individuals.\n",
      "Human: give examples and references\n",
      "Assistant: Sure! Some notable tools and approaches mentioned in the provided context that utilize machine learning in genomic selection (GS) are:\n",
      "1. rrBLUP\n",
      "2. solGS\n",
      "3. GMStool\n",
      "4. BWGS\n",
      "5. BGLR\n",
      "6. GenSel\n",
      "7. GSelection\n",
      "8. lme4GS\n",
      "\n",
      "These tools and approaches are used to evaluate genomic prediction accuracy and make genomic selection usable. For example, Cerioli et al. used the LSU500 marker set to achieve predictive abilities in various rice trials, and Zhang et al. studied Fusarium head blight (FHB) traits in wheat using GS. Zhang et al. also assessed maize drought tolerance using GS.\n",
      "\n",
      "Reference:\n",
      "- José Crossa, Paulino Pérez-Rodríguez, Jaime Cuevas, Osval Montesinos-López, Diego Jarquín, Gustavo de los Campos, Juan Burgueño, Juan M. González-Camacho, Sergio Pérez-Elizalde, Yoseph Beyene, Susanne Dreisigacker, Ravi Singh, Xuecai Zhang, Manje Gowda, Manish Roorkiwal, Jessica Rutkoski, and Rajeev K. Varshney. \"Genomic Selection in Plant Breeding: Methods, Models, and Perspectives.\"\n",
      "Follow Up Input: anyone use the BLUP?\n",
      "Standalone question:\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "especially in cross pollinated plants (Heffner et al., 2009). The rate\n",
      "of declination of selection response is more in GS than pedigree\n",
      "based selection, which can be minimized through the addition of\n",
      "new markers to the model (Nakaya and Isobe, 2012). However,\n",
      "the cost of implementation of GS is more than that of the\n",
      "traditional breeding program.\n",
      "The choice of models is an important factor in implementing\n",
      "GS, and several parametric and non-parametric genomic\n",
      "prediction models are available for this purpose. One of the\n",
      "most common and widely used parametric genomic selection\n",
      "model is the best linear unbiased prediction (BLUP). It is a\n",
      "mixed model– based whole-genome regression approach that is\n",
      "used to estimate the marker effects, and the same has been\n",
      "successfully applied to predict complex traits (Habier et al.,\n",
      "2009, 2013; de los Campos et al., 2013 ). In general, it was\n",
      "observed that the performance of parametric models found to be\n",
      "\n",
      "parents are selected based on their breeding values with high reliability, a faster\n",
      "genetic progress is expected in the resulting population. For this reason, the process\n",
      "of estimating breeding values is of paramount importance in any breeding program.\n",
      "There are several methods to estimate genomic estimated breeding values\n",
      "(GEBVs), but ﬁrst we will describe the best linear unbiased predictor (BLUP)\n",
      "method. When using the BLUP method to estimate the GEBVs, we need to use\n",
      "the mixed model equations ( 2.2) described above to estimate BLUEs and BLUPs.\n",
      "Using this equation ( 2.2) but depending on the form taken by the matrices Z and Σ,\n",
      "we can end up with the GBLUP method or the SNP-BLUP method to estimate the\n",
      "breeding values. First, we explain the GBLUP method, where we substitute Z and Σ\n",
      "matrices for the incidence matrix of genotypes and genomic relationship matrix\n",
      "(GRM) derived from allele frequencies calculated with one of the methods of\n",
      "52 2 Preprocessing Tools for Data Preparation\n",
      "\n",
      "Okeke UG, Akdemir D, Rabbi I, Kulakow P, Jannink JL (2017) Accuracies of univariate and\n",
      "multivariate genomic prediction models in African cassava. Genet Sel Evol 49(1):88\n",
      "Patterson HD, Thompson R (1971) Recovery of inter-block information when block sizes are\n",
      "unequal. Biometrika 58:545 –554\n",
      "Piepho HP, Möhring J, Melchinger AE, Büchse A (2008) BLUP for phenotypic selection in plant\n",
      "breeding and variety testing. Euphytica 161(1 –2):209–228\n",
      "Pinheiro JC, Bates DM (2000) Mixed-effects models in S and S-PLUS. Springer, New York\n",
      "Poland J, Endelman J, Dawson J, Rutkoski J, Wu S, Manes Y et al (2012) Genomic selection in\n",
      "wheat breeding using genotyping-by-sequencing. Plant Genome 5(3):103 –113\n",
      "Raudenbush SW, Bryk AS (2002) Hierarchical linear models: applications and data analysis\n",
      "methods. Sage Publications, Inc, Thousand Oaks, CA\n",
      "Rencher AC (2008) Linear models in statistics. Wiley, Hoboken, NJ\n",
      "Robinson GK (1991) That BLUP is a good thing: the estimation of random effects. Stat Sci 6\n",
      "\n",
      "108.100289\n",
      "Habier, D., Fernando, R. L., and Garrick, D. J. (2013). Genomic BLUP Decoded: a\n",
      "Look into the Black Box of Genomic Prediction.Genetics 194, 597– 607. doi:10.\n",
      "1534/GENETICS.113.152207\n",
      "Haile, T. A., Heidecker, T., Wright, D., Neupane, S., Ramsay, L., Vandenberg, A.,\n",
      "et al. (2020). Genomic Selection for Lentil Breeding: Empirical Evidence.Plant\n",
      "Genome 13, 1– 30. doi:10.1002/tpg2.20002\n",
      "Hayes, B. J., Bowman, P. J., Chamberlain, A. J., and Goddard, M. E. (2009). Invited\n",
      "Review: Genomic Selection in Dairy Cattle: Progress and Challenges.J. Dairy\n",
      "Sci. 92, 433– 443. doi:10.3168/JDS.2008-1646\n",
      "Heffner, E. L., Jannink, J.-L., Sorrells, M. E., Heffner, E. L., Sorrells, M. E., Univ, C.,\n",
      "et al. (2011). Genomic Selection Accuracy Using Multifamily Prediction Models\n",
      "in a Wheat Breeding Program.The Plant Genome 4, 65– 75. doi:10.3835/\n",
      "plantgenome.2010.12.0029\n",
      "Heffner, E. L., Lorenz, A. J., Jannink, J. L., and Sorrells, M. E. (2010). Plant Breeding\n",
      "Human: Have you or anyone else utilized the BLUP method in genomic selection?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "Yes, the BLUP (best linear unbiased prediction) method is a commonly used\n",
      "parametric genomic selection model. It has been successfully applied to estimate\n",
      "breeding values and predict complex traits in various breeding programs. Some of\n",
      "the references provided also discuss the application of BLUP in genomic\n",
      "selection (Habier et al., 2009, 2013; de los Campos et al., 2013).\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mGiven the following conversation and a follow up question, rephrase the follow up question to be a standalone question, in its original language.\n",
      "\n",
      "Chat History:\n",
      "\n",
      "Human: hello\n",
      "Assistant: Hello! How can I assist you today?\n",
      "Human: what is genomic breeding\n",
      "Assistant: Genomic breeding, also known as genomic selection (GS), is a modern plant breeding methodology that involves using molecular markers and genomic information to select the best candidates for breeding with high precision. By analyzing the genetic markers and whole-genome information of individuals, breeders can identify specific genes associated with desirable traits. This allows for accurate and efficient selection of individuals with the best genomic profiles for further breeding, ultimately accelerating the rate of genetic gain in crops and improving agronomic performance across different environmental conditions.\n",
      "Human: what is the disadvantages\n",
      "Assistant: Some of the disadvantages of genomic selection in breeding programs include factors like the cost of technology, potential inefficiency in chromosome doubling, limitations in selecting for basic traits during line development, and challenges in predicting accuracy influenced by various factors such as training and breeding population size, genetic diversity, genotype-environment interactions, density of markers, and genetic relationships between populations.\n",
      "Human: advantages?\n",
      "Assistant: Some advantages of genomic breeding in plant breeding programs include:\n",
      "\n",
      "1. **Precision**: Genomic tools allow for precise selection of desirable traits and prediction of breeding outcomes.\n",
      "   \n",
      "2. **Efficiency**: Genomic breeding methods enable more efficient selection and breeding processes, saving time and resources.\n",
      "   \n",
      "3. **Targeted Approach**: Genomic tools facilitate targeted genetic modifications, improving the accuracy of breeding programs.\n",
      "   \n",
      "4. **Identifying Favorable Alleles**: Genomic tools help in identifying existing favorable alleles of small effect that may have been overlooked previously.\n",
      "   \n",
      "5. **Accelerating Breeding Programs**: Genomic resources accelerate breeding programs by enabling the identification of genes associated with desirable traits.\n",
      "   \n",
      "6. **Enhancing Agricultural Productivity**: By improving crop yield, disease resistance, and stress tolerance, genomic breeding enhances agricultural productivity and sustainability.\n",
      "\n",
      "These advantages ultimately contribute to the development of new plant varieties with improved traits in a more precise, efficient, and targeted manner.\n",
      "Human: what should we pay attentation when doing that\n",
      "Assistant: When implementing genomic breeding in plant breeding programs, it is important to pay attention to the following factors:\n",
      "\n",
      "1. Genotyping cost\n",
      "2. Clear guidelines on where in the breeding program genomic selection (GS) could be efficiently applied\n",
      "3. Sufficient number of lines or animals in the reference (training) population\n",
      "4. Sufficient number of SNPs in the panels\n",
      "5. The homogeneity of individuals within the reference population (plants or animals)\n",
      "\n",
      "These factors are key to ensuring the success and efficiency of genomic selection in plant breeding programs.\n",
      "Human: what is the machine learning in GS\n",
      "Assistant: Machine learning plays a key role in genomic selection (GS) by helping to improve the methodology for genomic prediction accuracy and making GS usable. Machine learning methods are essential for evaluating genomic prediction accuracy, training models with reference populations, and predicting traits in non-phenotyped individuals. These methods are crucial in selecting the best candidates early in the breeding process, especially when genotypic information is limited or noisy. Therefore, machine learning tools contribute to enhancing the potential of GS by enabling more efficient and precise selection of candidate individuals.\n",
      "Human: give examples and references\n",
      "Assistant: Sure! Some notable tools and approaches mentioned in the provided context that utilize machine learning in genomic selection (GS) are:\n",
      "1. rrBLUP\n",
      "2. solGS\n",
      "3. GMStool\n",
      "4. BWGS\n",
      "5. BGLR\n",
      "6. GenSel\n",
      "7. GSelection\n",
      "8. lme4GS\n",
      "\n",
      "These tools and approaches are used to evaluate genomic prediction accuracy and make genomic selection usable. For example, Cerioli et al. used the LSU500 marker set to achieve predictive abilities in various rice trials, and Zhang et al. studied Fusarium head blight (FHB) traits in wheat using GS. Zhang et al. also assessed maize drought tolerance using GS.\n",
      "\n",
      "Reference:\n",
      "- José Crossa, Paulino Pérez-Rodríguez, Jaime Cuevas, Osval Montesinos-López, Diego Jarquín, Gustavo de los Campos, Juan Burgueño, Juan M. González-Camacho, Sergio Pérez-Elizalde, Yoseph Beyene, Susanne Dreisigacker, Ravi Singh, Xuecai Zhang, Manje Gowda, Manish Roorkiwal, Jessica Rutkoski, and Rajeev K. Varshney. \"Genomic Selection in Plant Breeding: Methods, Models, and Perspectives.\"\n",
      "Human: anyone use the BLUP?\n",
      "Assistant: Yes, the BLUP (best linear unbiased prediction) method is a commonly used parametric genomic selection model. It has been successfully applied to estimate breeding values and predict complex traits in various breeding programs. Some of the references provided also discuss the application of BLUP in genomic selection (Habier et al., 2009, 2013; de los Campos et al., 2013).\n",
      "Follow Up Input: how BLUP works\n",
      "Standalone question:\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "parents are selected based on their breeding values with high reliability, a faster\n",
      "genetic progress is expected in the resulting population. For this reason, the process\n",
      "of estimating breeding values is of paramount importance in any breeding program.\n",
      "There are several methods to estimate genomic estimated breeding values\n",
      "(GEBVs), but ﬁrst we will describe the best linear unbiased predictor (BLUP)\n",
      "method. When using the BLUP method to estimate the GEBVs, we need to use\n",
      "the mixed model equations ( 2.2) described above to estimate BLUEs and BLUPs.\n",
      "Using this equation ( 2.2) but depending on the form taken by the matrices Z and Σ,\n",
      "we can end up with the GBLUP method or the SNP-BLUP method to estimate the\n",
      "breeding values. First, we explain the GBLUP method, where we substitute Z and Σ\n",
      "matrices for the incidence matrix of genotypes and genomic relationship matrix\n",
      "(GRM) derived from allele frequencies calculated with one of the methods of\n",
      "52 2 Preprocessing Tools for Data Preparation\n",
      "\n",
      "inferences about model parameters are obtained on the basis of\n",
      "posterior distributions of marker effects. There are several\n",
      "variants of Bayesian models for genomic prediction such as\n",
      "Bayes A, Bayes B, Bayes Cπ, and Bayes Dπ (Meuwissen et al.,\n",
      "2001; Habier et al., 2011) and other derivatives, e.g., Bayesian\n",
      "LASSO and Bayesian ridge regression (BRR). Besides the marker-\n",
      "based models, the best linear unbiased prediction (BLUP)\n",
      "(Henderson et al., 1959) is one of the most commonly used\n",
      "genomic prediction methods. There are many variants of BLUP\n",
      "available for this purpose, e.g., genomic BLUP (GBLUP), single-\n",
      "step GBLUP (ssGBLUP), ridge regression BLUP (RRBLUP), and\n",
      "Frontiers in Genetics | www.frontiersin.org February 2022 | Volume 13 | Article 8321533\n",
      "Budhlakoti et al. A Comprehensive Review on Genomic Selection\n",
      "\n",
      "prediction of genotypic values, which uses line mean heritability and random\n",
      "genotypic effects that are structured by pedigree relations, can be generalized to a\n",
      "similar linear mixed with genotypic effects structured by marker relationships\n",
      "between genotypes. In that case, we move from classical best linear unbiased pre-\n",
      "dictions, or BLUPs, to genome-enabled BLUPs, or G-BLUPs.\n",
      "The G-BLUP has become the workhorse of genomic prediction in animal and\n",
      "plant breeding. For computational reasons, whenever the number of genotypes is\n",
      "less than the number of markers, a mixed model formulation that structures relation-\n",
      "ships between genotypes from marker pro ﬁles is more attractive than a formulation\n",
      "inserting a design matrix containing the SNPs itself. The latter formulation is a form\n",
      "of Ridge regression, and its predictions are called RR-BLUPs. G-BLUP and RR-\n",
      "BLUP can be motivated from a shrinkage perspective within a mixed model context,\n",
      "\n",
      "and is robust to extreme correlations among the predic -\n",
      "tors [46]. A Bayesian method to estimate marker effects is \n",
      "provided by Bayesian models like Bayes A, Bayes B, Bayes \n",
      "C, and Bayes D [56]. Genomic BLUP (GBLUP), one of the \n",
      "best linear unbiased prediction (BLUP) techniques, uses \n",
      "genomic relationships to estimate genetic values [131]. \n",
      "Ridge regression (RR), best linear unbiased prediction \n",
      "(BLUP), least absolute shrinkage and selection operator \n",
      "(LASSO), support vector machine (SVM), artificial neu -\n",
      "ral network (ANN), and random forest (RF) are examples \n",
      "of approaches based on single trait genomic selection \n",
      "(STGS). The multi-trait genomic selection technique \n",
      "(MTGS), which increases prediction accuracy for numer-\n",
      "ous characteristics, uses multivariate regression with \n",
      "covariance estimation (MRCE) and multivariate LASSO \n",
      "(MLASSO). A number of software tools and packages \n",
      "have been created to evaluate genomic prediction accu -\n",
      "Human: Can you explain how the BLUP (best linear unbiased prediction) method works in genomic selection?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "In genomic selection, the BLUP method works by estimating Genomic Estimated\n",
      "Breeding Values (GEBVs) using a linear mixed model. The BLUP method uses Mixed\n",
      "Model Equations to estimate Best Linear Unbiased Predictions (BLUPs) by\n",
      "incorporating genomic information. It involves substituting the incidence matrix\n",
      "of genotypes and the Genomic Relationship Matrix (GRM) derived from allele\n",
      "frequencies into the model. The BLUP method helps in predicting genotypic values\n",
      "by considering random genotypic effects structured by marker relationships,\n",
      "moving from classical BLUPs to genome-enabled BLUPs (G-BLUPs). G-BLUP is a\n",
      "widely used method in animal and plant breeding for genomic prediction,\n",
      "especially when the number of genotypes is less than the number of markers\n",
      "available. Additionally, there are variants of BLUP such as Genomic BLUP\n",
      "(GBLUP), Single-Step GBLUP (ssGBLUP), and Ridge Regression BLUP (RRBLUP) that\n",
      "offer different advantages in genomic selection.\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mGiven the following conversation and a follow up question, rephrase the follow up question to be a standalone question, in its original language.\n",
      "\n",
      "Chat History:\n",
      "\n",
      "Human: hello\n",
      "Assistant: Hello! How can I assist you today?\n",
      "Human: what is genomic breeding\n",
      "Assistant: Genomic breeding, also known as genomic selection (GS), is a modern plant breeding methodology that involves using molecular markers and genomic information to select the best candidates for breeding with high precision. By analyzing the genetic markers and whole-genome information of individuals, breeders can identify specific genes associated with desirable traits. This allows for accurate and efficient selection of individuals with the best genomic profiles for further breeding, ultimately accelerating the rate of genetic gain in crops and improving agronomic performance across different environmental conditions.\n",
      "Human: what is the disadvantages\n",
      "Assistant: Some of the disadvantages of genomic selection in breeding programs include factors like the cost of technology, potential inefficiency in chromosome doubling, limitations in selecting for basic traits during line development, and challenges in predicting accuracy influenced by various factors such as training and breeding population size, genetic diversity, genotype-environment interactions, density of markers, and genetic relationships between populations.\n",
      "Human: advantages?\n",
      "Assistant: Some advantages of genomic breeding in plant breeding programs include:\n",
      "\n",
      "1. **Precision**: Genomic tools allow for precise selection of desirable traits and prediction of breeding outcomes.\n",
      "   \n",
      "2. **Efficiency**: Genomic breeding methods enable more efficient selection and breeding processes, saving time and resources.\n",
      "   \n",
      "3. **Targeted Approach**: Genomic tools facilitate targeted genetic modifications, improving the accuracy of breeding programs.\n",
      "   \n",
      "4. **Identifying Favorable Alleles**: Genomic tools help in identifying existing favorable alleles of small effect that may have been overlooked previously.\n",
      "   \n",
      "5. **Accelerating Breeding Programs**: Genomic resources accelerate breeding programs by enabling the identification of genes associated with desirable traits.\n",
      "   \n",
      "6. **Enhancing Agricultural Productivity**: By improving crop yield, disease resistance, and stress tolerance, genomic breeding enhances agricultural productivity and sustainability.\n",
      "\n",
      "These advantages ultimately contribute to the development of new plant varieties with improved traits in a more precise, efficient, and targeted manner.\n",
      "Human: what should we pay attentation when doing that\n",
      "Assistant: When implementing genomic breeding in plant breeding programs, it is important to pay attention to the following factors:\n",
      "\n",
      "1. Genotyping cost\n",
      "2. Clear guidelines on where in the breeding program genomic selection (GS) could be efficiently applied\n",
      "3. Sufficient number of lines or animals in the reference (training) population\n",
      "4. Sufficient number of SNPs in the panels\n",
      "5. The homogeneity of individuals within the reference population (plants or animals)\n",
      "\n",
      "These factors are key to ensuring the success and efficiency of genomic selection in plant breeding programs.\n",
      "Human: what is the machine learning in GS\n",
      "Assistant: Machine learning plays a key role in genomic selection (GS) by helping to improve the methodology for genomic prediction accuracy and making GS usable. Machine learning methods are essential for evaluating genomic prediction accuracy, training models with reference populations, and predicting traits in non-phenotyped individuals. These methods are crucial in selecting the best candidates early in the breeding process, especially when genotypic information is limited or noisy. Therefore, machine learning tools contribute to enhancing the potential of GS by enabling more efficient and precise selection of candidate individuals.\n",
      "Human: give examples and references\n",
      "Assistant: Sure! Some notable tools and approaches mentioned in the provided context that utilize machine learning in genomic selection (GS) are:\n",
      "1. rrBLUP\n",
      "2. solGS\n",
      "3. GMStool\n",
      "4. BWGS\n",
      "5. BGLR\n",
      "6. GenSel\n",
      "7. GSelection\n",
      "8. lme4GS\n",
      "\n",
      "These tools and approaches are used to evaluate genomic prediction accuracy and make genomic selection usable. For example, Cerioli et al. used the LSU500 marker set to achieve predictive abilities in various rice trials, and Zhang et al. studied Fusarium head blight (FHB) traits in wheat using GS. Zhang et al. also assessed maize drought tolerance using GS.\n",
      "\n",
      "Reference:\n",
      "- José Crossa, Paulino Pérez-Rodríguez, Jaime Cuevas, Osval Montesinos-López, Diego Jarquín, Gustavo de los Campos, Juan Burgueño, Juan M. González-Camacho, Sergio Pérez-Elizalde, Yoseph Beyene, Susanne Dreisigacker, Ravi Singh, Xuecai Zhang, Manje Gowda, Manish Roorkiwal, Jessica Rutkoski, and Rajeev K. Varshney. \"Genomic Selection in Plant Breeding: Methods, Models, and Perspectives.\"\n",
      "Human: anyone use the BLUP?\n",
      "Assistant: Yes, the BLUP (best linear unbiased prediction) method is a commonly used parametric genomic selection model. It has been successfully applied to estimate breeding values and predict complex traits in various breeding programs. Some of the references provided also discuss the application of BLUP in genomic selection (Habier et al., 2009, 2013; de los Campos et al., 2013).\n",
      "Human: how BLUP works\n",
      "Assistant: In genomic selection, the BLUP method works by estimating Genomic Estimated Breeding Values (GEBVs) using a linear mixed model. The BLUP method uses Mixed Model Equations to estimate Best Linear Unbiased Predictions (BLUPs) by incorporating genomic information. It involves substituting the incidence matrix of genotypes and the Genomic Relationship Matrix (GRM) derived from allele frequencies into the model. The BLUP method helps in predicting genotypic values by considering random genotypic effects structured by marker relationships, moving from classical BLUPs to genome-enabled BLUPs (G-BLUPs). G-BLUP is a widely used method in animal and plant breeding for genomic prediction, especially when the number of genotypes is less than the number of markers available. Additionally, there are variants of BLUP such as Genomic BLUP (GBLUP), Single-Step GBLUP (ssGBLUP), and Ridge Regression BLUP (RRBLUP) that offer different advantages in genomic selection.\n",
      "Follow Up Input: how Bayer doing\n",
      "Standalone question:\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new StuffDocumentsChain chain...\u001b[0m\n",
      "\n",
      "\n",
      "\u001b[1m> Entering new LLMChain chain...\u001b[0m\n",
      "Prompt after formatting:\n",
      "\u001b[32;1m\u001b[1;3mSystem: Use the following pieces of context to answer the user's question. \n",
      "If you don't know the answer, just say that you don't know, don't try to make up an answer.\n",
      "----------------\n",
      "Are currently available GS models\n",
      "capable of providing the most accu-\n",
      "rate predictions?\n",
      "How can GS accelerate the ﬂow of\n",
      "favorable alleles directly from the gene\n",
      "bank to breeding programs?\n",
      "Are we ready to deploy GS in pre-\n",
      "breeding and germplasm-enhance-\n",
      "ment programs?\n",
      "Is it possible to combine GS and pedi-\n",
      "gree selection with HTP to accelerate\n",
      "genetic gains and save resources\n",
      "when developing lines during early-\n",
      "generation testing?\n",
      "Trends in Plant Science, November 2017, Vol. 22, No. 11 973\n",
      "\n",
      "GBLUP GBS Breeding population GY 0.12 –0.34 Sun et al. (2019)\n",
      "GBLUP and IBCF:MTME (item-based collaborativeﬁltering: multi-trait multi-environment)\n",
      "Illumina\n",
      "® 90 K Winter wheat lines GY -0.21 to 0.42 Lozada and Carter\n",
      "(2020)\n",
      "GBLUP and BRR In ﬁnium iSelect 9 K Germplasm Leaf rust resistance (LRR) 0.35 Daetwyler et al. (2014)\n",
      "Stem rust resistance (SRR) 0.27Yellow rust resistance (YRR) 0.44\n",
      "RR DArT Breeding lines Fusarium head blight (FHB) resistance 0.006 –0.463 Rutkoski et al. (2012)\n",
      "RKHS 0.118–0.575\n",
      "RF Deoxynivalenol (DON) resistance\n",
      "Bayesian LASSO and multiple linear regressionRRBLUP Illumina Inﬁnium 9 K and 90 K Winter wheat breeding lines FHB resistance 0.6 Mirdita et al. (2015)\n",
      "Bayes Cπ and RKHS Septoria leaf blotch resistance 0.5\n",
      "RRBLUP GBS Winter wheat breeding lines Powdery mildew resistance 0.60 Sarinelli et al. (2019)\n",
      "GY 0.64Test weight 0.71\n",
      "RKHS and GBLUP GBS Lines from International Bread Wheat Screening\n",
      "Nursery\n",
      "LRR Seedling 0.31 –0.74 Juliana et al. (2017)\n",
      "Adult 0.12 –0.56\n",
      "\n",
      "is a newly developed variety that is resistant. Suppose B1 and B2 are two chemicals used \n",
      "to control the pest or disease and B0 is “untreated”—no chemical applied. Finally, the \n",
      "response, Y, is some index of the plant’s health—or, more likely, the mean health of plants \n",
      "in the plot that forms the unit of replication.\n",
      "Now, imagine that under current practice, producers grow variety A0 and control the \n",
      "disease or pest using B2. However, there is accumulating evidence that using B2 may be a \n",
      "problem—for example, an environmental or health hazard. Suppose that B1 is alleged to \n",
      "be safer, although its effectiveness has not been shown. The objective of the study would \n",
      "then be to determine if we can obtain acceptable performance from the susceptible variety, \n",
      "A0, using B1, or, better yet, obtain equivalent performance from the resistant variety, A1, \n",
      "without using any chemical at all (B0).\n",
      "In this case, it makes sense to disregard the factorial structure of the treatments alto-\n",
      "\n",
      "F., Beyene, Y., Gowda, M., Gardner, K., et al.(2023b). Optimizing\n",
      "Sparse Testing for Genomic Prediction of Plant Breeding Crops.\n",
      "Genes 14:927. https://doi.org/10.3390/genes14040927.\n",
      "Molecular Plant 17, 552–578, April 1 2024ª 2024 The Author. 573\n",
      "Genomic selection in plant breeding Molecular Plant\n",
      "Human: How is Bayer performing currently?\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "\n",
      "\u001b[1m> Finished chain.\u001b[0m\n",
      "I don't have information on the current performance of Bayer.\n",
      "Exiting the conversation.\n"
     ]
    }
   ],
   "source": [
    "import textwrap\n",
    "\n",
    "# Initialize the chat model\n",
    "llm = ChatOpenAI()\n",
    "\n",
    "# Create a memory object to store the conversation history\n",
    "memory = ConversationBufferMemory(memory_key=\"chat_history\", return_messages=True)\n",
    "\n",
    "# Create a conversational retrieval chain by combining the chat model, retriever, and memory\n",
    "conversation = ConversationalRetrievalChain.from_llm(\n",
    "    llm, retriever=retriever, memory=memory, verbose=True\n",
    ")\n",
    "\n",
    "while True:\n",
    "    user_input = input(\"> \")  # Get user input\n",
    "    if user_input.lower() == \"exit\":  # Check if the user wants to exit\n",
    "        print(\"Exiting the conversation.\")\n",
    "        break  # Exit the loop\n",
    "    result = conversation.invoke(user_input)  # Invoke the conversation chain with the user input\n",
    "    wrapped_result = textwrap.fill(result[\"answer\"], width=80)\n",
    "    print(wrapped_result)\n",
    "    #print(\"\\n\".join(result[\"answer\"][i:i+80] for i in range(0, len(result[\"answer\"]), 80)))  # Print the chatbot's response with word wrapping"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
